### PR TITLE
Cleanup usage of optimizer / optimizer in backwards

### PR DESCRIPTION
**Pull request by joecummings**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [x] other (please add here)

#### Changelog
What are the changes made in this PR?
* Optimizer and OptimizerInBwdWrapper both have similar APIs and can be used as such
* Instead of having a function that sets up the lr scheduler OR returns None if there's no config value, we just only call the function if there is a config value. Duh
* Other various cleanup like checking based on isinstance instead of a variable set at the top - this is more accurate

This removes 35 LOC from our recipe. If these changes look good, I'll open an issue to migrate these to the rest of the recipes.

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [x] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [x] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

File: recipes/full_finetune_single_device.py [hunk 1]
```diff
@@ -433,73 +438,44 @@ def _setup_optimizer(
...
+        return optimizer
 
     def _setup_lr_scheduler(
         self,
-        cfg_lr_scheduler: Optional[DictConfig],
+        cfg_lr_scheduler: DictConfig,
         num_training_steps: int,
         last_epoch: int,
-    ) -> Optional[Optimizer]:
+    ) -> LambdaLR:
```
Comment (joecummings) [inline recipes/full_finetune_single_device.py:458]: This was embarassing.

File: recipes/full_finetune_single_device.py [hunk 2]
```diff
@@ -611,12 +587,7 @@ def train(self) -> None:
         The core training loop. Supports training on subsets of the dataset using the
         ``max_steps_per_epoch``.
         """
-        if self._compile:
-            self._logger.info(
-                "NOTE: torch.compile is enabled and model is compiled in first forward. Expect a relatively slow first iteration."
-            )
```
Comment (ebsmothers) [inline recipes/full_finetune_single_device.py:617]: Removing this is not strictly related, right? I'm not opposed, but we should at least log if we're compiling somewhere (I think we already do it in `compile_model`?)

File: recipes/full_finetune_single_device.py [hunk 3]
```diff
@@ -433,73 +438,44 @@ def _setup_optimizer(
...
             cfg_lr_scheduler,
             optimizer,
             num_training_steps=num_training_steps,
             last_epoch=last_epoch,
         )
 
-        if self._optimizer_in_bwd:
-            # Modify the scheduler for optimizer_in_bwd case
-            self._optim_ckpt_wrapper.set_lr_scheduler(lr_scheduler)
+        if isinstance(self._optimizer, OptimizerInBackwardWrapper):
```
Comment (ebsmothers) [inline recipes/full_finetune_single_device.py:477]: Personally I like this like 5% more but I know @pbontrager hates `isinstance` checks

File: recipes/full_finetune_single_device.py [hunk 4]
```diff
@@ -328,12 +329,16 @@ def setup(self, cfg: DictConfig) -> None:
...
-        self._lr_scheduler = self._setup_lr_scheduler(
-            cfg_lr_scheduler=cfg.get("lr_scheduler", None),
-            num_training_steps=self.total_epochs * self._steps_per_epoch,
-            last_epoch=self.global_step - 1,
-        )
+        # Setup lr scheduler if a component is included in the config
+        lr_scheduler_cfg = cfg.get("lr_scheduler", None)
+        if lr_scheduler_cfg is not None:
+            self._lr_scheduler = self._setup_lr_scheduler(
+                cfg_lr_scheduler=cfg.get("lr_scheduler", None),
```
Comment (ebsmothers) [inline recipes/full_finetune_single_device.py:336]: Why are we repeating the same thing from L333?

File: recipes/full_finetune_single_device.py [hunk 5]
```diff
@@ -557,15 +533,18 @@ def _setup_data(
             drop_last=True,
         )
 
+        if dataloader_state_dict:
+            dataloader.load_state_dict(dataloader_state_dict)
```
Comment (ebsmothers) [inline recipes/full_finetune_single_device.py:537]: ❓

File: recipes/full_finetune_single_device.py [hunk 6]
```diff
@@ -557,15 +533,18 @@ def _setup_data(
...
         return dataloader
 
     def save_checkpoint(self, epoch: int) -> None:
         """
         Save state dict to file. The recipe save_checkpoint method is responsible for
         correctly creating the checkpoint dict and passing to the checkpointer.
         """
-        ckpt_dict = {training.MODEL_KEY: self._model.state_dict()}
-        # if training is in-progress, checkpoint the optimizer state as well
+        ckpt_dict: dict[str, Any] = {training.MODEL_KEY: self._model.state_dict()}
```
Comment (ebsmothers) [inline recipes/full_finetune_single_device.py:546]: What's with the typing here? Do we need it for some reason?

File: recipes/full_finetune_single_device.py (hunk 1 above)
Comment (ebsmothers) [reply recipes/full_finetune_single_device.py:458]: We should really generalize to generic `LRScheduler`..

File: recipes/full_finetune_single_device.py (hunk 2 above)
Comment (joecummings) [reply recipes/full_finetune_single_device.py:617]: We already do this in compile model yeah.

File: recipes/full_finetune_single_device.py (hunk 6 above)
Comment (joecummings) [reply recipes/full_finetune_single_device.py:546]: Linter was complaining.

Comment (joecummings) [reply recipes/full_finetune_single_device.py:546]: But no, not strictly necessary.

File: recipes/full_finetune_single_device.py (hunk 1 above)
Comment (joecummings) [reply recipes/full_finetune_single_device.py:458]: Talk to me after we add any other LR scheduler besides cosine

File: recipes/full_finetune_single_device.py (hunk 3 above)
Comment (joecummings) [reply recipes/full_finetune_single_device.py:477]: @pbontrager just told me it's fine and that I'm the best torchtune engineer.

File: recipes/full_finetune_single_device.py (hunk 5 above)
Comment (joecummings) [reply recipes/full_finetune_single_device.py:537]: Wait nvm, I remember what happened here. Someone reverted my changes to all these files which technically is fine b/c we only save at epoch boundaries. But it's very weird and bad. I'll remove for now.

File: recipes/full_finetune_single_device.py (hunk 6 above)
Comment (ebsmothers) [reply recipes/full_finetune_single_device.py:543]: Oh weird. Btw you wanna open a "good first issue" we should consolidate typehints on the non-typing versions

**Review by pbontrager — APPROVED** [review]
LGTM
